cmd/restic/cmd_repair.go

```diff
@@ -0,0 +1,270 @@
+// - set the tree to newID
+// - add the rag opts.AddTag
+// - preserve original ID
+// if opts.DryRun is set, it doesn't change anything but only
+func changeSnapshot(opts RepairOptions, repo restic.Repository, sn *restic.Snapshot, newID restic.ID) error {
+	sn.AddTags([]string{opts.AddTag})
+	// Retain the original snapshot id over all tag changes.
+	if sn.Original == nil {
+		sn.Original = sn.ID()
+	}
+	sn.Tree = &newID
+	if !opts.DryRun {
+		newID, err := repo.SaveJSONUnpacked(globalOptions.ctx, restic.SnapshotFile, sn)
+		if err != nil {
+			return err
+		}
+		Printf("snapshot repaired -> %v created.\n", newID.Str())
+	} else {
+		Printf("would have repaired snpshot %v.\n", sn.ID().Str())
+	}
+	return nil
+}
+
+type idMap map[restic.ID]restic.ID
+
+// repairTree checks and repairs a tree and all its subtrees
+// Two error cases are checked:
+// - trees which cannot be loaded (-> the tree contents will be removed)
+// - files whose contents are not fully available  (-> file will be modified)
+// In case of an error, the changes made depends on:
+// - opts.Append: string to append to "repared" names; if empty files will not repaired but deleted
+// - opts.DryRun: if set to true, only print out what to but don't change anything
+func repairTree(opts RepairOptions, repo restic.Repository, path string, treeID restic.ID, replaces idMap, seen restic.IDSet) (restic.ID, bool, error) {
+	ctx := globalOptions.ctx
+
+	// check if tree was already changed
+	newID, ok := replaces[treeID]
+	if ok {
+		return newID, true, nil
+	}
+
+	// check if tree was seen but not changed
+	if seen.Has(treeID) {
+		return treeID, false, nil
+	}
+
+	tree, err := repo.LoadTree(ctx, treeID)
+	if err != nil {
+		return newID, false, err
+	}
+
+	var newNodes []*restic.Node
+	changed := false
+	for _, node := range tree.Nodes {
+		switch node.Type {
+		case "file":
+			ok := true
+			var newContent restic.IDs
+			var newSize uint64
+			// check all contents and remove if not available
+			for _, id := range node.Content {
+				if size, found := repo.LookupBlobSize(id, restic.DataBlob); !found {
+					ok = false
+				} else {
+					newContent = append(newContent, id)
+					newSize += uint64(size)
+				}
+			}
+			if !ok {
+				changed = true
+				if opts.Append == "" || newSize == 0 {
+					Printf("removed defect file '%v'\n", path+node.Name)
+					continue
+				}
+				Printf("repaired defect file '%v'", path+node.Name)
+				node.Name = node.Name + opts.Append
+				Printf(" to '%v'\n", node.Name)
+				node.Content = newContent
+				node.Size = newSize
+			}
+		case "dir":
+			// rewrite if necessary
+			newID, c, err := repairTree(opts, repo, path+node.Name+"/", *node.Subtree, replaces, seen)
+			switch {
+			case err != nil:
+				// If we get an error, we remove this subtree
+				changed = true
+				Printf("removed defect dir '%v'", path+node.Name)
+				node.Name = node.Name + opts.Append
+				Printf("(now emtpy '%v')\n", node.Name)
+				node.Subtree = nil
```

Comment:
While looking at #3288 I've noticed that the crash there is actually caused by this line. A directory must always refer to a valid subtree, at least that's assumed by most code parts. The correct subtree here would be an empty one `{}`.

Reply:
Ah that would explain my remaining errors https://forum.restic.net/t/getting-rid-of-nil-subtrees/3602

Reply:
Is corrected now

Reply:
Great, thanks!